worlds/tits_the_3rd/client/memory_io.py

```diff
@@ -273,3 +274,102 @@ def read_flag(self, flag_number: int):
         data = self._read_bytes(flag_byte_offset, 1)
         flag_value = (data >> flag_bit) & 1
         return bool(flag_value)
+
+    def get_scena_offset(self, scena_id: bytes):
+        """Get and verify a cached scena offset. Find and cache if it does not exist"""
+        if scena_id not in self.scena_offsets:
+            address = self.tits_the_3rd_mem.pattern_scan_all(scena_id)
+            if not address:
+                logger.error("Could not find give item function. Is the game running?")
+                return None
+            self.scena_offsets[scena_id] = address
+
+        if self.tits_the_3rd_mem.read_bytes(self.scena_offsets[scena_id], 2) != b"AP":
+            address = self.tits_the_3rd_mem.pattern_scan_all(scena_id)
+            if not address:
+                logger.error("Could not find give item function. Is the game running?")
+                return None
+            self.scena_offsets[scena_id] = address
+
+        return self.scena_offsets[scena_id]
+
+    def give_item(self, item_id: int, amount: int):
+        """Function to give an item to the player in game"""
+        address = self.get_scena_offset(b"AP Item Receive Notification")
+        if address is None:
+            logger.error("Could not find give item function. Is the game running?")
```

Comment:
Nit: this will double log (we already log in the get_scena_offset function). This is also the case for the calls below.